### PR TITLE
Page translation implementation

### PR DESCRIPTION
**Pull request by Mirageinvo**

### CODE REVIEW

File: memory/includes/common/mem.hpp [hunk 1]
```diff
@@ -8,6 +8,34 @@ constexpr uint64_t SHIFT_KB = 10ULL;
 constexpr uint64_t SHIFT_MB = 20ULL;
 constexpr uint64_t SHIFT_GB = 30ULL;
 
+template <int8_t shift, typename T = uint32_t>
```
Comment (PozhiloyPumba) [inline memory/includes/common/mem.hpp:11]: why this functions if we have the same in common/bitops.h

File: memory/virtual_mem.cpp [hunk 2]
```diff
@@ -1,10 +1,11 @@
 #include <iostream>
 #include "virtual_mem.hpp"
+#include "stdlib.h"
```
Comment (PozhiloyPumba) [inline memory/virtual_mem.cpp:3]: Is it needed?

File: memory/virtual_mem.cpp [hunk 3]
```diff
@@ -124,20 +79,46 @@ std::vector<uint8_t> VirtualMem::LoadByteSequence(uintptr_t addr, uint64_t lengt
...
-    return phys_addr;
+    auto paddr3 = page_translation_[vpn3];
+    auto paddr2 = paddr3[vpn2];
+    auto paddr1 = paddr2[vpn1];
+    auto paddr0 = paddr1[vpn0];
+    if (paddr0 == 0) {
+        uint64_t pageNum = ram_->GetEmptyPageNumber();
+        ram_->InitPage(pageNum);
+        paddr0 = pageNum;
+        page_translation_[vpn3][vpn2][vpn1][vpn0] = pageNum;
```
Comment (PozhiloyPumba) [inline memory/virtual_mem.cpp:92]: Maybe you can get reference to this earlier and don't do 4 times addresation in arrays again again?

**Review by PozhiloyPumba — CHANGES_REQUESTED** [review]

File: memory/includes/common/mem.hpp (hunk 1 above)
Comment (Mirageinvo) [reply memory/includes/common/mem.hpp]: Removed, thanks

File: memory/virtual_mem.cpp (hunk 2 above)
Comment (Mirageinvo) [reply memory/virtual_mem.cpp]: Removed

File: memory/includes/phys_mem.hpp [hunk 4]
```diff
@@ -19,18 +16,21 @@ class PhysMem final {
...
+    uint8_t *GetMemPointer() const;
 
 private:
     PhysMem(uint64_t size);
-    ~PhysMem() = default;
+    ~PhysMem();
 
     uint64_t total_size_;
-    std::vector<Page> pages_;
+    // TODO(Mirageinvo): return Page class?
```
Comment (MrYar0s) [inline memory/includes/phys_mem.hpp:31]: Remove TODO, it's better solution then std::vector<Page>

Comment (Mirageinvo) [reply memory/includes/phys_mem.hpp]: Done

**Review by MrYar0s — APPROVED** [review]

**Pull request merged**